### ISSUE TITLE
Remove API Lock

### ISSUE THREAD

**Issue by kunstmusik** [csound/csound#859]
Remove use of API_lock and replace with Disruptor-style message queue (ring buffer with pre-allocated messages).  The result will be that the API will be lock-free/wait-free with no runtime memory allocations, but with the addition of max messages allowed (pre-allocated at startup but before runtime).  

The proposed design is discussed in this mailing list thread:

https://listserv.heanet.ie/cgi-bin/wa?A2=ind1710&L=CSOUND-DEV&D=0&P=27480

**Comment by vlazzarini** [comment]
How do you do this? Do you pass the relevant API function pointer and its parameters in the circular buffer? Do we need to maintain a table of API functions that has all of them? It's not quite clear to me.

**Comment by kunstmusik** [comment]
The buffer would have pre-allocated message data structures.  We could make them hold whatever we want.  For example, in Pink I have this:

https://github.com/kunstmusik/pink/blob/master/src/main/pink/node/Message.java
https://github.com/kunstmusik/pink/blob/master/src/main/pink/node/MessageBuffer.java

It's Java, but hopefully makes sense.  The Message object has values for a keyword and a function.  In our case, we could have something like:

```
struct APIMessage {
   int messageType;
   void* arguments[16];  // maximum 16 arguments
}

The messageType would then be used in a big case statement on the reading side, and arguments could be stored by reference in the arguments array.  

At least, this was what I was imagining.

**Comment by nwhetsell** [comment]
It seems likely that this issue is related to https://github.com/csound/csound/issues/772. If so, here is what I did in csound-api as a workaround; it’s pretty much exactly what @kunstmusik describes, but with a lock-free queue instead of a ring buffer:

https://github.com/nwhetsell/csound-api/blob/90e750a16e60a00154abbd608de92038e6bbb8c1/src/csound-api.cc#L515-L604

I’m pretty sure this is also the idea behind [csound_threaded.hpp](https://github.com/csound/csound/blob/develop/include/csound_threaded.hpp).

As @kunstmusik notes in the message on the mailing list, this would likely be a major change to the Csound API, especially since many API functions return status codes.

**Comment by vlazzarini** [comment]
I understand. I was thinking of a function table rather than switch, but it's the same idea. My worry is that this will have to be well maintained and in sync with any additions to the API.

It's the right thing to do, we just need to make sure it is kept in sync.

**Comment by kunstmusik** [comment]
@nwhetsell Yes, this is related and addresses the same issue but avoids any runtime allocations.

I had not thought about return codes.  Perhaps what we can do is keep the lock, but add new API methods, such as:

csoundReadScoreAsync
csoundInputMessageAsync
csoundCompileOrchAsync

that would use this message system. If all of these functions follow the same signature (CSOUND, char*), we could make it where the message is something like:

```
struct AsyncMessage {
    void (apiFunc*)(CSOUND*, char*);
    char* text;
}
```

and then we'd get around the need to sync the API.  (Or, make it such that the table and message system is not exposed via the API and used only internally.)

**Comment by vlazzarini** [comment]
I'd just like to point out that #772, while related to the question here, is not completely due to it. In that case, there is another thread involved and some amount of buffering, which depending on the platform and buffer size, is also prone to jitter. 

Also with Java (and this can be at play in the Blue issue), there are regular blocks caused by garbage collection. If csoundPerformKsmps is called by a java thread, it will be prone to jitter.
If CsoundPerformThread is used, then that's OK.

Also @kunstmusik, what you proposed (in your latest comment) is already implemented in CsoundPerformanceThread. It might be better to think of a system that applies more widely to the API, to replace the current threadsafe implementation.

**Comment by kunstmusik** [comment]
@vlazzarini I don't believe GC is in play here as the behavior is consistent: larger ksmps causes longer blocking and vice versa.  This is because PerformKsmps locks the API_lock for its entire duration and causes contention with the other API function calls. 

CsoundPerformanceThread is also not an option for me (at least, as it is today) as I run Csound in a Java thread and do operations before and after each PerformKsmps call. (The operations need to be done synchronously with the Csound engine.)  Also, it is allocating/deleting at runtime, which the ring buffer would avoid.  

Or, am I missing something?

**Comment by vlazzarini** [comment]
I am saying this because I observed the jitter on Android with Java calls, when working on the beathealth code. Moving to placing the performKsmps calls in the C thread (OpenSL) fixed the problem.

In my checks there were blocks of around 40ms at times, regularly happening. We did extensively an analysis of this. So I am just saying it in case you still experience similar issues.

Also note that what you observed has not been reported elsewhere (I would have thought that Cabbage users at least would have noticed it, and maybe with CsoundQt too).

Having said that, I agree we should aim to get rid of the locks.

**Comment by kunstmusik** [comment]
Hmm, I see, I will have to do some testing here then as it could be
something else and I had only been testing with Blue.  I will leave it for
the moment though and come back once we start looking into the lock
removal.

**Comment by gogins** [comment]
What Steven proposes is not really the same as CsoundPerformanceThread. The
queues used are different. Steven's proposal does not allocate for the
queue during performance, it is pre-allocated.

Regards,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
That's just a detail, but in principle it does what he described.

In any case, I've gone ahead and implemented a messaging queue system in Top/threadsafe.c:

1. Not rolled out yet. Relevant API functions still use the lock
2. Of course, not yet tested.
3. Does not allow (yet) for a complete removal of the lock. There remains some uses
with compilation (csound_orc_compile.c, main.c), killing instances (insert.c), cleaning up (musmon.c), which we might be able to remove in a similar way.

@kunstmusik I've added the code for you to review. See what you think. If we're confident we might just roll it in before 6.10.

**Comment by nwhetsell** [comment]
@vlazzarini This might be better managed as a feature branch.

For what it’s worth:

* The queue size is defined as a macro ([`API_MAX_QUEUE `](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L46)). I think it would be better if this was configurable, so that the queue size could be set using a command line option, for example.

* I could be missing something, but the name of the message struct, [`message_queue_t`](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L51), seems to me to be a misnomer. `message_t` might be more appropriate.

* It might be better to use an identifier with the [message type enum](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L44) rather than an anonymous enum (and update the [`message` member](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L49) of the message struct accordingly). (Also, `messageType` might be a better name for the `message` member.)

* Restricting the size of all message arguments to [128 bytes](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L50) strikes me as a bit severe. I can easily imagine needing to call `csoundReadScore` with much larger scores.

* The updates of the queue’s [read](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L96) and [write](https://github.com/csound/csound/blob/27daaf2b87fe7d1158087d734a0df31744abe4fd/Top/threadsafe.c#L62) indexes seem off. (If you increment the write index when it equals `API_MAX_QUEUE`, then the index will refer to something beyond the end of the queue.)

**Comment by vlazzarini** [comment]
arguments to strings and arrays are to be passed as pointers, so 128 seems OK to me.  Thanks for the note on the updates, I'll check the code to fix it.

**Comment by kunstmusik** [comment]
@vlazzarini I did a build with MinGW and tested with Blue. Performance is snappy, even getting rid of some other glitching I had when not using Blue's Virtual Keyboard.  

On the other hand, this does not compile with Visual Studio, mainly due to variable-length arrays which VS does not support:

```
1>  threadsafe.c
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(157): warning C4024: 'csoundKillInstance': different types for formal and actual parameter 3
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2133: 'args': unknown size
```

Also, I think the use of atomics in threadsafe.c may not be correct.  Two things: 

1. I think we should use CAS in a loop to calculate then set the value
2. We need VS-compatible atomic code (and probably need to review our codebase wherever we have HAVE_ATOMIC_BUILTIN and implement VS-compatible code)

I find the much better performance from Blue very nice though. I did try UDP server from Csound and found there are still delays with ksmps > 1, but I don't know if that is related to this work or not.

**Comment by fsateler** [comment]
@vlazzarini Directly addressing locations and casting to non-char types (like [here](https://github.com/csound/csound/blob/cb96b4e8f179be941398856d10587ae16e381249/Top/threadsafe.c#L99)) is not safe, because it can result in unaligned access. It can produce a SIGBUS in some architectures, and it causes slowdowns in others. The safe way is to do the reverse of what you did in the `*_enqueue` methods: memcpy from the array into a target variable. Eg, something like (untested):

```c
      case SCORE_EVENT:
	{
          char* i = csound->msg_queue[rp].args;
	  char type = *i;
          i+=1;
	  MYFLT *pfields;
          long numFields;
          memcpy(&pfields, i, sizeof(MYFLT*));
          i+=1;
          memcpy(&numFields, i, sizeof(long));
	  csoundScoreEventInternal(csound, type, pfields, numFields);
	}
```

**Comment by gogins** [comment]
Simplify your life and don't use a preallocated queue or a lockfree queue.
Just use c++11 which has already been debugged, thank you. The performance
hit is minuscule compared to what else is going on in Csound. If you don't
want burps when the queue grows, then resize it to something configurable
and large before running Csound. Guaranteed to be much easier this way. You
are trying to kill a mosquito with a cannon.

Best,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
Which functions did you remove the lock and tested?

On 12 Oct 2017, at 16:34, Steven Yi <notifications@github.com<mailto:notifications@github.com>> wrote:


@vlazzarini<https://github.com/vlazzarini> I did a build with MinGW and tested with Blue. Performance is snappy, even getting rid of some other glitching I had when not using Blue's Virtual Keyboard.

On the other hand, this does not compile with Visual Studio, mainly due to variable-length arrays which VS does not support:

1>  threadsafe.c
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(157): warning C4024: 'csoundKillInstance': different types for formal and actual parameter 3
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(172): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(179): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(189): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(198): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(208): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(219): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(233): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(245): error C2133: 'args': unknown size
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2057: expected constant expression
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2466: cannot allocate an array of constant size 0
1>C:\Users\stevenyi\work\csound\csound\Top\threadsafe.c(254): error C2133: 'args': unknown size


Also, I think the use of atomics in threadsafe.c may not be correct. Two things:

  1.  I think we should use CAS in a loop to calculate then set the value
  2.  We need VS-compatible atomic code (and probably need to review our codebase wherever we have HAVE_ATOMIC_BUILTIN and implement VS-compatible code)

I find the much better performance from Blue very nice though. I did try UDP server from Csound and found there are still delays with ksmps > 1, but I don't know if that is related to this work or not.

**Comment by kunstmusik** [comment]
@vlazzarini I did not remove anything for testing; the code is valid for MinGW's GCC, but not valid for Visual Studio.  I tested with the MinGW/GCC build.

**Comment by vlazzarini** [comment]
@kunstmusik But the lockfree code has not been rolled out. The functions in threadsafe.c need to be changed for that. I've done some basic testing just to verify the code is working as it should be.

@fsateler thanks for this. I'd have expected that the array data would always be contiguous. I'll see what I can do to address it.

**Comment by kunstmusik** [comment]
@vlazzarini - UGH, my mistake: I forgot I had done some settings experiments last night here and that was what I was seeing. (-b, -B and ksmps changes affected Blue Virtual Keyboard usage).  

For these changes, is it the goal to replace existing API functions, or add these async options? I.e., should testing be to write new code for API?  (BTW: This would probably be a good unit test candidate)

**Comment by vlazzarini** [comment]
@kunstmusik  for testing: modify the API functions as marked in threadsafe.c 
Then we need to decide what to do.

I've modified the code so that it should be better in visual C (I was using C11 stuff it probably doesn't like).

**Comment by vlazzarini** [comment]
Provided it works, there is a couple of decisions we need to make

1) whether we will replace the lock-based functions with the lock-free code
2) what we will do with the functions that return a status code.
    of these there are two types: a) ReadScore and Event functions;
    b) Compilation and Instance killing
  
  with a) we might get away just returning OK as I don't think hosts
check the results.
  with b) I think I can re-arrange the code so that it returns a code
  in the compilation case, because the locked part of the code is
 only the merging part, after compilation passed. So I can isolate
it to be asynchronous and only be inserted if compilation was successful.
The instance killing is maybe trickier, but I think this function has not been
used a lot, so we might get away with always returning OK.

Alternatively, we just provide async versions of these functions. This means
we have to keep the locks in place still, and that may not be ideal.

**Comment by nwhetsell** [comment]
There’s one more function that returns a value, but not a status code: [`csoundEvalCode`](https://csound.github.io/docs/api/group__PERFORMANCE.html#ga50cd39501bafba67115bf91c516cd1b8).

**Comment by vlazzarini** [comment]
That's a harder one to tackle.

**Comment by fsateler** [comment]
@vlazzarini 

> I'd have expected that the array data would always be contiguous. I'll see what I can do to address it.

The problem is not that the array is not contiguous. It's that doing eg, `*(double*)args` will not result in the same assembly instruction as `*(char*)args` , as the first will read 8 bytes, and the second will read just one. Some architectures require that for using the 'read 8 bytes' instruction the address specified be a multiple of 8. These crash with SIGBUS. Other architectures detect that the address is not a multiple of 8, and correct for it. These have a performance cost. There's more information

**Comment by vlazzarini** [comment]
@fsateler yes I figured it out that alignment would be problematic. Anyway, I replaced the assignments with memcpys.

**Comment by vlazzarini** [comment]
This is done now. It needs testing (I've only tested a little bit), and we have to make a decision on
how to roll it out:

1. Do we want to replace the existing API functions? I have solved the problem of returned
values for all but the score/event functions. We can then remove the API locks completely from
the codebase.
2. Do we want to provide *Async() versions? This is safer. I have supplied these but have not
added to the interface (csound.h). We could do this with a view to mainstreaming them in the
next major version. However, we would still have the locks in place.

**Comment by nwhetsell** [comment]
For what it’s worth, my preference would be to provide new `*Async` functions, not to replace the existing API functions, especially if the new code lacks test coverage.

**Comment by fggp** [comment]
If you change the API functions signature in csound.h, I'll have to change
ctcsound.py and csnd.go accordingly. Can you tell me when you feel that
csound.h is fixed and stable, so that I can proceed.

Regards.

François

2017-10-13 14:47 GMT+02:00 vlazzarini <notifications@github.com>:

**Comment by vlazzarini** [comment]
@fggp No change in signature planned in any of the two options.

**Comment by stekyne** [comment]
I don't think you should change the API at all. Just change how it is implemented underneath. The client API should function the same as before, i.e. block until it receives a return value.

**Comment by vlazzarini** [comment]
I am not proposing to change the API, but its implementation. But of course if we are to remove the locks, there is a certain amount of underlying behaviour change.

**Comment by vlazzarini** [comment]
OK, since we have not had any further views, I am going to add new API functions instead of replacing the old ones. That means we will postpone removing the locks completely to 7.

**Comment by kunstmusik** [comment]
We will still need to update the new async code for VS build, as well as
review the atomics changes.  (Shouldn't affect anything except UDP server
since everything else would use the existing synchronous API.)

**Comment by vlazzarini** [comment]
Do you have an idea of what changes VS needs? Sorry I can't help you there.
The atomic operations are fine for the design I used, at least as far as I can see. 
We could have used CAS, but then the queue design would be different.

**Comment by kunstmusik** [comment]
I think we'd be able to use the _InterlockedXXX functions:

https://docs.microsoft.com/en-us/cpp/intrinsics/intrinsics-available-on-all-architectures

but I'm not certain as it's first time for me to lookup atomic functions for VS.

I think for message_enqueue, a CAS loop is going to be necessary as the wp could have changed by a separate thread between the read at the top and the write near the bottom.  At least, that's on a quick look; I'll try to take a look at this more carefully tomorrow.

**Comment by vlazzarini** [comment]
We should not need the CAS, because we pick the value of wp only once, and use it. Even if it increments in between, it doesn't matter, since we'll deal with that in the next call. By design, wp is always ahead of rp, unless there is an overflow (which is unlikely).

**Comment by vlazzarini** [comment]
To be clear: wp is only changed by one thread and read by both. We protect the reading and writing.

**Comment by stekyne** [comment]
Yeah, you need to use those Interlocked functions. This is already sort of done in the cs_new_dispatch.c file, around ~307.

#if defined(_MSC_VER)
#define ATOMIC_CAS(x,current,new)  InterlockedCompareExchange(x, current, new)

Maybe this needs to be standardised across the lib in sysdep.h?
